### PR TITLE
@W-16045041 Reference implementation for Inventory Calculator

### PR DESCRIPTION
**Pull request by l-wei791**

### CODE REVIEW

**Comment by pplat23** [comment]
@l-wei791 
1. Could you please add somebody from your team for a review first? To access how accurate and meaningful Apex implementations and tests are?
2. I see this PR contains implementation of `commerce_inventory.CommerceInventoryService` which is [InventoryService Extension](https://developer.salesforce.com/docs/commerce/salesforce-commerce/guide/available-extensions.html#inventory-service). Could you please also add implementation of `CartExtension.InventoryCartCalculator` which is [Inventory Cart Calculator Extension](https://developer.salesforce.com/docs/commerce/salesforce-commerce/guide/available-extensions.html#inventory-cart-calculator)

File: commerce/domain/inventory/CommerceInventoryServiceSample.cls [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            responseItem.setReservedAtLocationId(item.getReservedAtLocationId());
+            responseItem.setItemReservationSourceId(item.getItemReservationSourceId());
+            responseItem.setProductId(item.getProductId());
+            responseItems.add(responseItem);
+        }
+
+        response.setItems(responseItems);
+
+        if (upsertReservationRequest.getItems().size() != response.getItems().size()) {
+            throw new NoDataFoundException();
```
Comment (jasonmoodysf) [inline commerce/domain/inventory/CommerceInventoryServiceSample.cls:24]: Please change the error type.  It should be something reflecting a data Conflict.  

Also please assign the boolean expression to a variable to make it super clear what the error condition is: 

eg: 
boolean reservationConflict = upsertReservationRequest.getItems().size() != response.getItems().size(); 
if (reservationConflict) {
 throw new DataConflict("Expected " + x + " got y") 
}
If you need to create a custom execption type: https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_exception_custom.htm

File: commerce/domain/inventory/CommerceInventoryServiceSample.cls [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    public override commerce_inventory.InventoryLevelsResponse getInventoryLevel(commerce_inventory.InventoryLevelsRequest request) {
+        commerce_inventory.InventoryLevelsResponse response = new commerce_inventory.InventoryLevelsResponse();
+        Set<commerce_inventory.InventoryLevelsItemResponse> items = new Set<commerce_inventory.InventoryLevelsItemResponse>();
+        for(commerce_inventory.InventoryLevelsItemRequest item : request.getItemInventoryLevelRequests()) {
+            commerce_inventory.InventoryLevelsItemResponse itemResponse = new commerce_inventory.InventoryLevelsItemResponse();
+            itemResponse.setProductId(item.getProductId());
+            itemResponse.setLocationSourceId(item.getLocationSourceId());
+            itemResponse.setInventoryLocationSourceType('LocationGroup');
+            itemResponse.setOnHand(double.valueOf('10.0'));
```
Comment (jasonmoodysf) [inline commerce/domain/inventory/CommerceInventoryServiceSample.cls:54]: remove all hard coded values in this method.

File: commerce/domain/inventory/CommerceInventoryServiceSample.cls [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            itemResponse.setAvailableToFulfill(double.valueOf('10.0'));
+            itemResponse.setAvailableToOrder(double.valueOf('10.0'));
+            items.add(itemResponse);
+        }
+
+        response.setItemsInventoryLevels(items);
+
+        if (response.getItemsInventoryLevels().size() != request.getItemInventoryLevelRequests().size()) {
+            throw new NoDataFoundException();
```
Comment (jasonmoodysf) [inline commerce/domain/inventory/CommerceInventoryServiceSample.cls:63]: Move this code in to a shared method.  Fix it the same way as above.

**Review by jasonmoodysf — CHANGES_REQUESTED** [review]

File: commerce/domain/inventory/CommerceInventoryServiceSample.cls (hunk 2 above)
Comment (sf-afoliaco) [reply commerce/domain/inventory/CommerceInventoryServiceSample.cls:54]: @jasonmoodysf  what will the values be? this is sample inventory data.

**Comment by sf-afoliaco** [comment]
> @l-wei791
> 
> 1. Could you please add somebody from your team for a review first? To access how accurate and meaningful Apex implementations and tests are?
> 2. I see this PR contains implementation of `commerce_inventory.CommerceInventoryService` which is [InventoryService Extension](https://developer.salesforce.com/docs/commerce/salesforce-commerce/guide/available-extensions.html#inventory-service). Could you please also add implementation of `CartExtension.InventoryCartCalculator` which is [Inventory Cart Calculator Extension](https://developer.salesforce.com/docs/commerce/salesforce-commerce/guide/available-extensions.html#inventory-cart-calculator)

@pplat23 can you please add me to repo so I can contribute, so far I can't

**Pull request closed without merging**